### ISSUE TITLE
bug: Can't set individual request timeouts

### ISSUE THREAD

**Issue by amunra** [algesten/ureq#758]
# TL;DR

Timeout values don't override correctly between the `AgentBuilder` and `Request` APIs.

# Problem

This took us some time to debug.

We need to set an overall request timeout on a per-request basis.
What we found is that `Request::timeout` is ignored some times.

The issue seems to be that the API docs aren't quite accurate and the behaviour is less than desireable.

The `Request::timeout` docs "Sets overall timeout for the request, overriding agent’s configuration if any." imply that this API should be the one-stop shop for setting the timeout on the request.

Instead the `AgentBuilder::timeout_connection` default of 30 seconds takes priority if a connection is not already established.

# Suggested new behaviour

The `Request::timeout` should always be respected and override `AgentBuilder::timeout_connection`.
Moreover the precedence order here looks inverted: `AgentBuilder::timeout` should also override `AgentBuilder::timeout_connection` if set.

In other words, I suggest (pseudo-code):

```python
connection_timeout = min(
    agent.timeout_connection or 30s,
    agent.timeout or inf,
    request.timeout or inf)
```

The new `AgentBuilder::timeout_connection` docs would then read:
```
Timeout for the socket connection to be successful. If both this and `.timeout()` are both set, `.timeout()` takes precedence.

The default is 30 seconds.
```

This change would resolve the conflicting documentation and fix the feature so it becomes possible to set a per-request timeout.

**Comment by algesten** [comment]
Hi! This sounds similar to something we discussed before: https://github.com/algesten/ureq/issues/614

Is it the same?

**Comment by amunra** [comment]
> Hi! This sounds similar to something we discussed before: #614
> 
> Is it the same?

Yes, I believe it is. Closing and continuing there.

This bug is a duplicate of #614.
